fix(setup-checks): Ensure URL with webroot works

susnux opened the pull request:
* Resolves https://github.com/nextcloud/server/issues/47880

## Summary
We basically mock the way `URLGenerator::getAbsoluteURL` works, so we must make sure that the URL might already contain the webroot. Because `baseURL` and `cliURL` also contain the webroot we need to remove the webroot from the URL first.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

susnux committed:
fix(setup-checks): Ensure URL with webroot works

We basically mock the way `URLGenerator::getAbsoluteURL` works,
so we must make sure that the URL might already contain the webroot.
Because `baseURL` and `cliURL` also contain the webroot we need to remove
the webroot from the URL first.

Co-authored-by: Ferdinand Thiessen <opensource@fthiessen.de>
Co-authored-by: Daniel <mail@danielkesselberg.de>
Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
/backport to stable30

solracsf commented on apps/settings/lib/SetupChecks/CheckServerResponseTrait.php:
```diff
@@ -41,6 +41,15 @@ protected function serverConfigHelp(): string {
+			$url = '/' . $url;
+		}
+		// Similar to `getAbsoluteURL` of URLGenerator:
+		// The Nextcloud web root could be already be prepended.
```
```suggestion
		// The Nextcloud web root could already be prepended.
```

solracsf commented on lib/private/URLGenerator.php:
```diff
@@ -264,7 +264,7 @@ public function getAbsoluteURL(string $url): string {
 			return rtrim($this->config->getSystemValueString('overwrite.cli.url'), '/') . '/' . ltrim($url, '/');
 		}
-		// The ownCloud web root can already be prepended.
+		// The Nextcloud web root could be already be prepended.
```
```suggestion
		// The Nextcloud web root could already be prepended.
```

kesselb commented:
https://github.com/nextcloud/server/issues/47880 => 29.0.6
https://github.com/nextcloud/server/pull/46255 => 29.0.7

The bug report is for 29.0.6, Come and I just finished a patch for 29.0.7 to resolve some issues with webroot.

We still need this patch and if yes, how's the case different?

susnux commented on apps/settings/lib/SetupChecks/CheckServerResponseTrait.php:
```diff
@@ -76,8 +85,8 @@ protected function getTestUrls(string $url, bool $removeWebroot): array {
 			}
-			$hosts[] = 'https://' . $host . $url;
-			$hosts[] = 'http://' . $host . $url;
+			$testUrls[] = $this->normalizeUrl("https://$host/$webroot", $webroot, $removeWebroot) . $url;
```
@kesselb see this: If you install into `/nextcloud/` then this will be wrong otherwise as the trusted domain does not contain the webroot.
Meaning we need to add the webroot here.

susnux replied:
(also those URLs were not added but only "hosts")

susnux commented on apps/settings/lib/SetupChecks/CheckServerResponseTrait.php:
```diff
@@ -41,6 +41,15 @@ protected function serverConfigHelp(): string {
+		}
+		// Similar to `getAbsoluteURL` of URLGenerator:
+		// The Nextcloud web root could already be prepended.
+		if (\OC::$WEBROOT !== '' && str_starts_with($url, \OC::$WEBROOT)) {
```
@kesselb This is needed because if you use the URLGenerator (e.g. the JavaScriptModules test), the resulting URL already contains the webroot.

Then you concat `cliUrl` or `baseUrl` with that url, but both of them also contain the webroot.

E.g. the URL is `/nextcloud/apps/settings/js/test.mjs` and the `cliUrl` is `https://example.com/nextcloud`

susnux commented:
> We still need this patch and if yes, how's the case different?

See the inline comments I just added to point out current bugs we have with webroots.

kesselb commented:
Thanks @susnux :+1: 

1) I made a mistake in normalizeUrl. https://github.com/nextcloud/server/blob/897c341c5d894aa76ad1823f8f656f589515a903/apps/settings/lib/SetupChecks/CheckServerResponseTrait.php#L91-L93 That should be ` substr($url, 0, -strlen($webroot));`  to remove the webroot (right now it does extract the webroot :see_no_evil:)

2) We did not adjust the javascript checks. They runHead method should also accept a "removeWebroot" parameter and forward it to the runRequest method.

What do you think?

susnux commented:
@kesselb good point, fixed that as well and added proper unit tests to make sure we cover all cases.

> They runHead method should also accept a "removeWebroot" parameter

I do not think so, the tests **should** contain the webroot the request is not agains the root like with well-known URLs.
I think developers should not need to care about if the URLGenerator adds the webroot or not, you should just put in your expected URL and decide: Is this a root request or a Nextcloud request?
Please see my added unit tests :)

kesselb commented on apps/settings/lib/SetupChecks/CheckServerResponseTrait.php:
```diff
@@ -38,66 +38,69 @@ protected function serverConfigHelp(): string {
 	 */
 	protected function getTestUrls(string $url, bool $removeWebroot): array {
-		$testUrls = [];
+		if (!str_starts_with($url, '/')) {
+			$url = '/' . $url;
+		}
```
```suggestion
		$url = '/' . ltrim($url, '/');
```

Maybe?

kesselb commented:
```
  'overwrite.cli.url' => 'https://server.internal/testcloud',
  'overwritewebroot' => '/testcloud',
```

- Run `occ setupchecks` 
- Set a breakpoint in testUrls
- Check JavaScriptModules 

![image](https://github.com/user-attachments/assets/390aa710-3c5c-4662-b39c-a3740aab06b4)

The generated URL contains the webroot twice: `https://server.internal/testcloud/testcloud/apps/settings/js/esm-test.mjs` 

I don't know whether the webroot should start with a slash or not. The referenced bug is using a leading slash.

kesselb commented:
Too many forward slashes for me :see_no_evil:

kesselb commented:
/backport to stable30

kesselb commented:
/backport to stable29

come-nc commented on apps/settings/lib/SetupChecks/DataDirectoryProtected.php:
```diff
@@ -41,8 +41,7 @@ public function getName(): string {
 
 	public function run(): SetupResult {
 		$datadir = str_replace(\OC::$SERVERROOT . '/', '', $this->config->getSystemValue('datadirectory', ''));
```
```suggestion
		$datadir = str_replace(\OC::$SERVERROOT . '/', '', $this->config->getSystemValueString('datadirectory', ''));
```

kesselb commented on apps/settings/lib/SetupChecks/CheckServerResponseTrait.php:
```diff
@@ -38,66 +38,67 @@ protected function serverConfigHelp(): string {
 
 		$webroot = rtrim($this->urlGenerator->getWebroot(), '/');
+		// Similar to `getAbsoluteURL` of URLGenerator:
+		// The Nextcloud web root could already be prepended.
```
Good point :+1: 

@susnux explained his idea/approach here: https://github.com/nextcloud/server/pull/47883#issuecomment-2343440098

I also think we should have one way for the URLs and not accept different formats. 

Example JavaScriptModules: https://github.com/nextcloud/server/blob/8feef455068533c18f7ab25e2274187733217a5e/apps/settings/lib/SetupChecks/JavaScriptModules.php#L43

- Is using the url generator and thus contains the webroot already.

Example WellKnownUrls: https://github.com/nextcloud/server/blob/8feef455068533c18f7ab25e2274187733217a5e/apps/settings/lib/SetupChecks/WellKnownUrls.php#L46-L51

- Not using the url generator.


1. We added the option to strip the webroot with https://github.com/nextcloud/server/pull/46255 for the wellknown use case to using the cli url (that includes the webroot) and remove the webroot for the root request. 
2. I think it's also okay to require that all urls, to test, contain the webroot if necessary. 

Nice would be to have only one way, to avoid further bewilderment ;)

susnux replied:
I refactored it like discussed:

- Always expect absolute path (so including the web-root if needed, meaning you can just use the output of the URLGenerator)
- Option to make the request a root request which removes the webroot from CLI URL and base URL.

susnux committed:
refactor(settings): `CheckServerResponseTrait` always expect absolute path

Co-authored-by: Ferdinand Thiessen <opensource@fthiessen.de>
Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

come-nc commented on apps/settings/lib/SetupChecks/CheckServerResponseTrait.php:
```diff
@@ -37,85 +37,93 @@ protected function serverConfigHelp(): string {
 	/**
 	 * Strip a trailing slash and remove the webroot if requested.
+	 * @param string $url The URL to normalize. Should be an absolute URL containing scheme, host and optionally web-root.
+	 * @param string $removeWebroot If set the web-root is removed from the URL and an absolute URL with only the scheme and host (optional port) is returned
```
```suggestion
	 * @param bool $removeWebroot If set the web-root is removed from the URL and an absolute URL with only the scheme and host (optional port) is returned
```

come-nc reviewed (approved):
I think fixing the wrong type in the phpdoc should fix psalm errors.
Nice job :+1:

come-nc commented:
phpunit tests also need to be adapted to change from getSystemValue to getSystemValueString
